File: libadb/src/main/java/io/github/muntashirakon/adb/android/AdbMdns.java
// SPDX-License-Identifier: GPL-3.0-or-later OR Apache-2.0

package io.github.muntashirakon.adb.android;

import android.content.Context;
import android.net.nsd.NsdManager;
import android.net.nsd.NsdServiceInfo;
import android.os.Build;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;
import androidx.annotation.StringDef;

import java.io.IOException;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.NetworkInterface;
import java.net.ServerSocket;
import java.net.SocketException;
import java.util.Collections;
import java.util.Objects;

/**
 * Automatic discovery of ADB daemons.
 */
// Copyright 2020 南宫雪珊
// Copyright 2022 Muntashir Al-Islam
// Based on https://android.googlesource.com/platform/packages/modules/adb/+/eddd2d3a386a83f5d1e14f87a318adef4c2f1a9d/adb_mdns.cpp
@RequiresApi(Build.VERSION_CODES.JELLY_BEAN)
public class AdbMdns {
    public static final String SERVICE_TYPE_ADB = "adb";
    public static final String SERVICE_TYPE_TLS_PAIRING = "adb-tls-pairing";
    public static final String SERVICE_TYPE_TLS_CONNECT = "adb-tls-connect";

    @StringDef({
            SERVICE_TYPE_ADB,
            SERVICE_TYPE_TLS_PAIRING,
            SERVICE_TYPE_TLS_CONNECT,
    })
    @Retention(RetentionPolicy.SOURCE)
    public @interface ServiceType {
    }

    public interface OnAdbDaemonDiscoveredListener {
        void onPortChanged(@Nullable InetAddress hostAddress, int port);
    }

    @NonNull
    private final Context mContext;
    @NonNull
    private final String mServiceType;
    @NonNull
    private final OnAdbDaemonDiscoveredListener mAdbDaemonDiscoveredListener;
    private final NsdManager.DiscoveryListener mDiscoveryListener;
    private final NsdManager mNsdManager;

    private boolean mRegistered;
    private boolean mRunning;
    @Nullable
    private String mServiceName;

    public AdbMdns(@NonNull Context context, @ServiceType @NonNull String serviceType,
                   @NonNull OnAdbDaemonDiscoveredListener portChangeListener) {
        mContext = Objects.requireNonNull(context);
        mServiceType = String.format("_%s._tcp", Objects.requireNonNull(serviceType));
        mAdbDaemonDiscoveredListener = Objects.requireNonNull(portChangeListener);
        mNsdManager = (NsdManager) context.getSystemService(Context.NSD_SERVICE);
        mDiscoveryListener = new DiscoveryListener(this);
    }

    public void start() {
        if (mRunning) return;
        mRunning = true;
        if (!mRegistered) {
            mNsdManager.discoverServices(mServiceType, NsdManager.PROTOCOL_DNS_SD, mDiscoveryListener);
        }
    }

    public void stop() {
        if (!mRunning) return;
        mRunning = false;
        if (mRegistered) {
            mNsdManager.stopServiceDiscovery(mDiscoveryListener);
        }
    }

    public boolean isRunning() {
        return mRunning;
    }

    private void onDiscoveryStart() {
        mRegistered = true;
    }

    private void onDiscoverStop() {
        mRegistered = false;
    }

    private void onServiceFound(NsdServiceInfo serviceInfo) {
        mNsdManager.resolveService(serviceInfo, new ResolveListener(this));
    }

    private void onServiceLost(NsdServiceInfo serviceInfo) {
        if (mServiceName != null && mServiceName.equals(serviceInfo.getServiceName())) {
            mAdbDaemonDiscoveredListener.onPortChanged(serviceInfo.getHost(), -1);
        }
    }

    private void onServiceResolved(NsdServiceInfo serviceInfo) {
        if (!mRunning) return;
        try {
            for (NetworkInterface networkInterface : Collections.list(NetworkInterface.getNetworkInterfaces())) {
                for (InetAddress inetAddress : Collections.list(networkInterface.getInetAddresses())) {
                    String inetHost = inetAddress.getHostAddress();
                    if (Objects.equals(inetHost, serviceInfo.getHost().getHostAddress())
                            && isPortAvailable(serviceInfo.getPort())) {
                        mServiceName = serviceInfo.getServiceName();
                        mAdbDaemonDiscoveredListener.onPortChanged(serviceInfo.getHost(), serviceInfo.getPort());
                    }
                }
            }
        } catch (SocketException e) {
            e.printStackTrace();
        }
    }

    private boolean isPortAvailable(int port) {
        try (ServerSocket socket = new ServerSocket()) {
            socket.bind(new InetSocketAddress(AndroidUtils.getHostIpAddress(mContext), port), 1);
            return false;
        } catch (IOException e) {
            return true;
        }
    }

    private static class DiscoveryListener implements NsdManager.DiscoveryListener {
        @NonNull
        private final AdbMdns mAdbMdns;

        private DiscoveryListener(@NonNull AdbMdns adbMdns) {
            mAdbMdns = adbMdns;
        }

        @Override
        public void onDiscoveryStarted(String serviceType) {
            mAdbMdns.onDiscoveryStart();
        }

        @Override
        public void onStartDiscoveryFailed(String serviceType, int errorCode) {
        }

        @Override
        public void onDiscoveryStopped(String serviceType) {
            mAdbMdns.onDiscoverStop();
        }

        @Override
        public void onStopDiscoveryFailed(String serviceType, int errorCode) {
        }

        @Override
        public void onServiceFound(NsdServiceInfo serviceInfo) {
            mAdbMdns.onServiceFound(serviceInfo);
        }

        @Override
        public void onServiceLost(NsdServiceInfo serviceInfo) {
            mAdbMdns.onServiceLost(serviceInfo);
        }
    }

    private static class ResolveListener implements NsdManager.ResolveListener {
        @NonNull
        private final AdbMdns mAdbMdns;

        private ResolveListener(@NonNull AdbMdns adbMdns) {
            mAdbMdns = adbMdns;
        }

        @Override
        public void onResolveFailed(NsdServiceInfo serviceInfo, int errorCode) {
        }

        @Override
        public void onServiceResolved(NsdServiceInfo serviceInfo) {
            mAdbMdns.onServiceResolved(serviceInfo);
        }
    }
}

File: app/src/main/java/com/drdisagree/colorblendr/utils/wifiadb/AdbMdns.kt
package com.drdisagree.colorblendr.utils.wifiadb

import android.content.Context
import android.net.nsd.NsdManager
import android.net.nsd.NsdServiceInfo
import android.os.Handler
import android.os.Looper
import android.util.Log
import java.io.IOException
import java.net.InetSocketAddress
import java.net.NetworkInterface
import java.net.ServerSocket
import java.util.Collections

/**
 * Class for managing ADB over mDNS (Multicast DNS) service discovery.
 *
 * @param context The application context.
 * @param callback The callback interface to handle detected ADB services.
 */
@Suppress("DEPRECATION")
class AdbMdns(context: Context, private val callback: AdbFoundCallback) {

    private val resolvedServices: MutableSet<String?> = HashSet()

    private val nsdManager = context.getSystemService(Context.NSD_SERVICE) as NsdManager
    private var running = false
    private var stopResolving = false
    private val handler: Handler = Handler(Looper.getMainLooper())

    private val pairingListener = DiscoveryListener(TLS_PAIRING)
    private val connectListener = DiscoveryListener(TLS_CONNECT)

    /**
     * Callback interface for handling detected ADB services.
     */
    interface AdbFoundCallback {
        /**
         * Called when a pairing code service is detected.
         *
         * @param ipAddress The IP address of the service.
         * @param port The port number of the service.
         */
        fun onPairingCodeDetected(ipAddress: String, port: Int)

        /**
         * Called when a connect code service is detected.
         *
         * @param ipAddress The IP address of the service.
         * @param port The port number of the service.
         */
        fun onConnectCodeDetected(ipAddress: String, port: Int)
    }

    /**
     * Starts the mDNS service discovery for ADB services.
     */
    fun start() {
        if (running) return

        try {
            nsdManager.discoverServices(TLS_PAIRING, NsdManager.PROTOCOL_DNS_SD, pairingListener)
            nsdManager.discoverServices(TLS_CONNECT, NsdManager.PROTOCOL_DNS_SD, connectListener)
            running = true
        } catch (e: Exception) {
            Log.e(TAG, "Error starting service discovery: ", e)
            running = false
        }
    }

    /**
     * Stops the mDNS service discovery for ADB services.
     */
    fun stop() {
        if (!running) return

        try {
            nsdManager.stopServiceDiscovery(pairingListener)
            nsdManager.stopServiceDiscovery(connectListener)
        } catch (e: Exception) {
            Log.e(TAG, "Error stopping service discovery: ", e)
        }

        resolvedServices.clear()
        handler.removeCallbacksAndMessages(null)
        stopResolving = false
        running = false
    }

    /**
     * Checks if the resolved service matches the current network.
     *
     * @param resolvedService The resolved service information.
     * @return True if the service matches the current network, false otherwise.
     */
    private fun isMatchingNetwork(resolvedService: NsdServiceInfo): Boolean {
        try {
            val interfaces = NetworkInterface.getNetworkInterfaces()
            for (networkInterface in Collections.list(interfaces)) {
                if (networkInterface.isUp()) {
                    for (inetAddress in Collections.list(networkInterface.inetAddresses)) {
                        if (resolvedService.host.hostAddress == inetAddress.hostAddress) {
                            return true
                        }
                    }
                }
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }
        return false
    }

    /**
     * Checks if the specified port is available.
     *
     * @param port The port number to check.
     * @return True if the port is available, false otherwise.
     */
    private fun isPortAvailable(port: Int): Boolean {
        try {
            ServerSocket().use { socket ->
                socket.bind(InetSocketAddress("127.0.0.1", port), 1)
                return false
            }
        } catch (_: IOException) {
            return true
        }
    }

    /**
     * Listener for mDNS service discovery events.
     *
     * @param serviceType The type of service to discover.
     */
    private inner class DiscoveryListener(
        private val serviceType: String
    ) : NsdManager.DiscoveryListener {

        override fun onDiscoveryStarted(serviceType: String?) {
            Log.v(TAG, "Discovery started: $serviceType")
        }

        override fun onStartDiscoveryFailed(serviceType: String?, errorCode: Int) {
            Log.v(TAG, "Start discovery failed: $serviceType, $errorCode")
        }

        override fun onDiscoveryStopped(serviceType: String?) {
            Log.v(TAG, "Discovery stopped: $serviceType")
        }

        override fun onStopDiscoveryFailed(serviceType: String?, errorCode: Int) {
            Log.v(TAG, "Stop discovery failed: $serviceType, $errorCode")
        }

        override fun onServiceFound(info: NsdServiceInfo) {
            if (!running) return

            val serviceKey = "${info.host}:${info.port}"

            if (!resolvedServices.contains(serviceKey)) {
                resolvedServices.add(serviceKey)
                nsdManager.resolveService(info, ResolveListener(serviceType))

                handler.removeCallbacksAndMessages(null)
                handler.postDelayed({ stopResolving = true }, 3 * 60 * 1000)
            } else if (!stopResolving) {
                nsdManager.resolveService(info, ResolveListener(serviceType))
            }
        }

        override fun onServiceLost(info: NsdServiceInfo?) {
            if (info == null || info.host == null) return

            val serviceKey = "${info.host.hostAddress}:${info.port}"
            resolvedServices.remove(serviceKey)
            Log.d(TAG, "Service lost: $serviceKey")
        }
    }

    /**
     * Listener for resolving mDNS services.
     *
     * @param serviceType The type of service being resolved.
     */
    private inner class ResolveListener(
        private val serviceType: String
    ) : NsdManager.ResolveListener {

        override fun onResolveFailed(nsdServiceInfo: NsdServiceInfo, errorCode: Int) {
            Log.v(
                TAG,
                "Resolve failed: " + nsdServiceInfo.serviceName + ", errorCode: " + errorCode
            )
        }

        override fun onServiceResolved(resolvedService: NsdServiceInfo) {
            if (running
                && isMatchingNetwork(resolvedService)
                && isPortAvailable(resolvedService.port)
            ) {
                val ipAddress = resolvedService.host.hostAddress
                val portNumber = resolvedService.port

                if (serviceType == TLS_PAIRING) {
                    callback.onPairingCodeDetected(ipAddress, portNumber)
                } else if (serviceType == TLS_CONNECT) {
                    callback.onConnectCodeDetected(ipAddress, portNumber)
                }
            }
        }
    }

    companion object {
        private const val TAG = "AdbMdns"
        const val TLS_CONNECT: String = "_adb-tls-connect._tcp"
        const val TLS_PAIRING: String = "_adb-tls-pairing._tcp"
    }
}